File: src/main/java/me/miki/shindo/management/remote/download/DownloadManager.java
package me.miki.shindo.management.remote.download;


import me.miki.shindo.Shindo;
import me.miki.shindo.management.file.FileManager;
import me.miki.shindo.management.remote.download.file.DownloadFile;
import me.miki.shindo.management.remote.download.file.DownloadZipFile;
import me.miki.shindo.utils.Multithreading;
import me.miki.shindo.utils.file.DirectoryUtils;
import me.miki.shindo.utils.file.FileExtractor;
import me.miki.shindo.utils.network.HttpUtils;

import java.io.File;
import java.util.ArrayList;

public class DownloadManager {


    private final ArrayList<DownloadFile> downloadFiles = new ArrayList<DownloadFile>();

    private boolean downloaded;

    public DownloadManager() {

        FileManager fileManager = Shindo.getInstance().getFileManager();

        downloaded = false;


        Multithreading.runAsync(this::startDownloads);
    }

    private void startDownloads() {

        for (DownloadFile df : downloadFiles) {

            if (!df.getOutputDir().exists()) {
                df.getOutputDir().mkdirs();
            }

            if (df instanceof DownloadZipFile) {

                DownloadZipFile dzf = (DownloadZipFile) df;

                if (DirectoryUtils.getDirectorySize(dzf.getOutputDir()) != dzf.getUnzippedSize()) {

                    File outputFile = new File(dzf.getOutputDir(), dzf.getFileName());

                    HttpUtils.downloadFile(dzf.getUrl(), outputFile);
                    FileExtractor.unzip(outputFile, dzf.getOutputDir());
                    outputFile.delete();
                }
            } else {

                File outputFile = new File(df.getOutputDir(), df.getFileName());

                if (outputFile.length() != df.getSize()) {
                    HttpUtils.downloadFile(df.getUrl(), outputFile);
                }
            }
        }

        checkFiles();
    }

    private void checkFiles() {

        for (DownloadFile df : downloadFiles) {

            if (df instanceof DownloadZipFile) {

                DownloadZipFile dzf = (DownloadZipFile) df;

                if (DirectoryUtils.getDirectorySize(dzf.getOutputDir()) != dzf.getUnzippedSize()) {
                    startDownloads();
                }
            } else {

                File outputFile = new File(df.getOutputDir(), df.getFileName());

                if (outputFile.length() != df.getSize()) {
                    startDownloads();
                }
            }
        }

        downloaded = true;
    }

    public boolean isDownloaded() {
        return downloaded;
    }
}

File: src/main/kotlin/me/miki/shindo/management/remote/download/DownloadManager.kt
package me.miki.shindo.management.remote.download

import me.miki.shindo.management.remote.download.file.DownloadFile
import me.miki.shindo.management.remote.download.file.DownloadZipFile
import me.miki.shindo.utils.concurrent.TaskExecutor
import me.miki.shindo.utils.concurrent.ThreadPoolType
import me.miki.shindo.utils.file.DirectoryUtils
import me.miki.shindo.utils.file.FileExtractor
import me.miki.shindo.utils.network.HttpUtils
import java.io.File

class DownloadManager {
    private val downloadFiles = ArrayList<DownloadFile>()
    private var downloaded = false

    init {
        TaskExecutor.runAsync(ThreadPoolType.NETWORK) { startDownloads() }
    }

    private fun startDownloads() {
        for (df in downloadFiles) {
            if (!df.outputDir.exists()) df.outputDir.mkdirs()
            when (df) {
                is DownloadZipFile -> {
                    val dzf = df
                    if (DirectoryUtils.getDirectorySize(dzf.outputDir) != dzf.unzippedSize) {
                        val outputFile = File(dzf.outputDir, dzf.fileName)
                        HttpUtils.downloadFile(dzf.url, outputFile)
                        FileExtractor.unzip(outputFile, dzf.outputDir)
                        outputFile.delete()
                    }
                }

                else -> {
                    val outputFile = File(df.outputDir, df.fileName)
                    if (outputFile.length() != df.size) HttpUtils.downloadFile(df.url, outputFile)
                }
            }
        }
        checkFiles()
    }

    private fun checkFiles() {
        for (df in downloadFiles) {
            when (df) {
                is DownloadZipFile -> {
                    val dzf = df
                    if (DirectoryUtils.getDirectorySize(dzf.outputDir) != dzf.unzippedSize) {
                        startDownloads()
                        return
                    }
                }

                else -> {
                    val outputFile = File(df.outputDir, df.fileName)
                    if (outputFile.length() != df.size) {
                        startDownloads()
                        return
                    }
                }
            }
        }
        downloaded = true
    }

    fun isDownloaded(): Boolean = downloaded
}
